src/System.Data.Odbc/tests/OdbcParameterTests.cs

```diff
@@ -0,0 +1,265 @@
+            Assert.Equal(str4000, outputStr);
+            Assert.Equal(str4000.Length * 2, inputLength); // since NVARCHAR takes 2 bytes per character
+            Assert.Equal(str4000.Length * 2, outputLength);
+
+            RunTestProcedure("NVARCHAR", 4000, str5000, out output, out inputLength, out outputLength);
+            outputStr = output as string;
+            Assert.Equal(str4000, outputStr);
+            Assert.Equal(str4000.Length * 2, inputLength);
+            Assert.Equal(str4000.Length * 2, outputLength);
+
+            RunTestProcedure("NVARCHAR", 4000, str4000, out output, out inputLength, out outputLength);
+            outputStr = output as string;
+            Assert.Equal(str4000, outputStr);
+            Assert.Equal(str4000.Length * 2, inputLength);
+            Assert.Equal(str4000.Length * 2, outputLength);
+
+            RunTestProcedure("NVARCHAR", 4000, str2000, out output, out inputLength, out outputLength);
+            outputStr = output as string;
+            Assert.Equal(str4000, outputStr);
+            Assert.Equal(str2000.Length * 2, inputLength);
+            Assert.Equal(str4000.Length * 2, outputLength);
+
+            RunTestProcedure("NVARCHAR", 4000, str1000, out output, out inputLength, out outputLength);
+            outputStr = output as string;
+            Assert.Equal(str2000, outputStr);
+            Assert.Equal(str1000.Length * 2, inputLength);
+            Assert.Equal(str2000.Length * 2, outputLength);
+        }
+        
+
+        private static void RunTestProcedure(string procDataType, int procDataSize, object v1, out object v2, out int v3, out int v4)
+        {
+            string removeExistingStoredProcSql =
+                "IF OBJECT_ID('testStoredProc', 'P') IS NOT NULL " +
+                    "DROP PROCEDURE testStoredProc;";
+
+            string createTestStoredProcSql =
+                "CREATE PROCEDURE testStoredProc (" +
+                    $"@v1 {procDataType}({procDataSize}), " +
+                    $"@v2 {procDataType}({procDataSize}) OUT, " +
+                    "@v3 INTEGER OUT, " +
+                    "@v4 INTEGER OUT) " +
+                "AS BEGIN " +
+                    "SET @v2 = @v1 + @v1; " +
+                    "SET @v3 = datalength(@v1); " +
+                    "SET @v4 = datalength(@v2); " +
+                "END;";
+
+            DataTestUtility.RunNonQuery(DataTestUtility.TcpConnStr, removeExistingStoredProcSql);
+            DataTestUtility.RunNonQuery(DataTestUtility.TcpConnStr, createTestStoredProcSql);
+
+            DbAccessor dbAccessUtil = new DbAccessor();
+            dbAccessUtil.connectSqlServer(DataTestUtility.OdbcConnStr);
+            dbAccessUtil.callProc("{ call testStoredProc(?,?,?,?) }", procDataType, procDataSize, v1, out v2, out v3, out v4);
+            dbAccessUtil.commit();
+            dbAccessUtil.disconnect();
+
+            DataTestUtility.RunNonQuery(DataTestUtility.TcpConnStr, removeExistingStoredProcSql);
```

Comment:
Put this cleanup in a try-finally, in case the test db access calls throw an exception.